### PR TITLE
Corner Kick Validation

### PR DESCRIPTION
**Pull request by jared-paul**
### Description

Adds simulated test validation functions to the corner kick play test

### Testing Done

No tests for the validation functions as of now

### Resolved Issues

#1608 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by jared-paul** [comment]
Write now I'm just getting the first and second robot found in world->getFriendlyRobots()->getAllRobots() and checking if they are involved in the play. My laptop is not the best so I can't tell which robots are actually supposed to be involved in the play lol. Because of this though I was able to try and look through why the test failed, but it was hard for me to pinpoint an actual reason. Maybe this is because I'm not too familiar with the output? @MathewMacDougall 

I will update this PR with unit tests for the validation functions once an initial review is done

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (jonathanlew) [inline src/software/simulated_tests/corner_kick_play_test.cpp]: need to remove this to run in CI

File: src/software/simulated_tests/defense_play_test.cpp
Comment (jonathanlew) [inline src/software/simulated_tests/defense_play_test.cpp]: same here

File: src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.cpp
Comment (jonathanlew) [inline src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.cpp]: can we just pass in the `Ball`?

File: src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.h
Comment (jonathanlew) [inline src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.h]: missing new line and javadocs

File: src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp
Comment (jonathanlew) [inline src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp]: I think we should pass in `Ball` here instead of `world_ptr`

**Review by jonathanlew — CHANGES_REQUESTED** [review]
looks like it's going in the right direction. I will run this on my computer

**Comment by jonathanlew** [comment]
> Write now I'm just getting the first and second robot found in world->getFriendlyRobots()->getAllRobots() and checking if they are involved in the play. My laptop is not the best so I can't tell which robots are actually supposed to be involved in the play lol. Because of this though I was able to try and look through why the test failed, but it was hard for me to pinpoint an actual reason. Maybe this is because I'm not too familiar with the output? @MathewMacDougall
> 
> I will update this PR with unit tests for the validation functions once an initial review is done

1. it goes bot 5 then bot 4
2. I changed it to bot5 then bot4 and it seems like robotHasPossession isn't working to detect possession. It might be better to make your own algorithm
3. friendlyScored works

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/corner_kick_play_test.cpp]: I know you said this is kinda hacky / temp right now, but you can get robot's from the Team by their ID with `getRobotById()`

Comment (MathewMacDougall) [inline src/software/simulated_tests/corner_kick_play_test.cpp]: Perhaps leave a TODO here linked to a ticket to also add a check for the robot kicking the ball? IIRC we were doing that later after some changes were made to how we detect kicks since it was non-blocking for now.

File: src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp]: Since `result` is an optional we should check if it actually contains a value before trying to get the value.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Thanks @jonathanlew for testing. Couple more comments, but looking good :+1:

**Comment by MathewMacDougall** [comment]
@jared-paul just curious do things just run so slowly they lag-out and the test ends before you can see everything? I assumed because the physics steps and visualizations were done synchronously even slower machines should be able to see the test, even if it was at a very slow FPS.

**Comment by jared-paul** [comment]
> @jared-paul just curious do things just run so slowly they lag-out and the test ends before you can see everything? I assumed because the physics steps and visualizations were done synchronously even slower machines should be able to see the test, even if it was at a very slow FPS.

It was just very laggy (very low FPS). I think one time it crashed too, but I think it was because I'm constantly running at 7/8 gb of RAM for some reason

**Comment by jared-paul** [comment]
> > Write now I'm just getting the first and second robot found in world->getFriendlyRobots()->getAllRobots() and checking if they are involved in the play. My laptop is not the best so I can't tell which robots are actually supposed to be involved in the play lol. Because of this though I was able to try and look through why the test failed, but it was hard for me to pinpoint an actual reason. Maybe this is because I'm not too familiar with the output? @MathewMacDougall
> > I will update this PR with unit tests for the validation functions once an initial review is done
> 
> 1. it goes bot 5 then bot 4
> 2. I changed it to bot5 then bot4 and it seems like robotHasPossession isn't working to detect possession. It might be better to make your own algorithm
> 3. friendlyScored works

Thank you for testing this, I will try and resolve my RAM issue (only happens in Ubuntu, in Windows my RAM is at a usable amount)

**Comment by MathewMacDougall** [comment]
> It was just very laggy (very low FPS). I think one time it crashed too, but I think it was because I'm constantly running at 7/8 gb of RAM for some reason

CLion can use a lot of ram to load symbols and stuff, and then takes more when it runs bazel builds. You could try closing your IDE and just running the test from the command line, which should take less RAM overall.

**Comment by jared-paul** [comment]
It seems as though the majority of time I run the test it goes Robot 5 -> Robot 4 -> Goal **BUT** one time I ran it it went Robot 5 -> Robot 3 -> Goal so... this could potentially be a **big** problem. Anyone have any thoughts? @MathewMacDougall @jonathanlew @akhilveeraghanta

**Comment by jonathanlew** [comment]
> It seems as though the majority of time I run the test it goes Robot 5 -> Robot 4 -> Goal **BUT** one time I ran it it went Robot 5 -> Robot 3 -> Goal so... this could potentially be a **big** problem. Anyone have any thoughts? @MathewMacDougall @jonathanlew @akhilveeraghanta

this would probably fix the remaining non-determinism but it's soft blocked on dynamic parameter stuff https://github.com/UBC-Thunderbots/Software/issues/766
For now, we can ignore it if it will still pass CI...

**Comment by MathewMacDougall** [comment]
Yup that's 99% due to non-determinism in the `PassGenerator` as Jon said. Since it works most of the time you could mark the bazel test as `flaky` which will cause it to re-run 3 times before declaring it a failure. That should help prevent random CI failures in the meantime (but should be fixed properly long-term).

https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes-tests

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (jared-paul) [reply src/software/simulated_tests/corner_kick_play_test.cpp]: resolved

File: src/software/simulated_tests/defense_play_test.cpp
Comment (jared-paul) [reply src/software/simulated_tests/defense_play_test.cpp]: resolved

File: src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.cpp
Comment (jared-paul) [reply src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.cpp]: I think we determined we can't because only the world_ptr get's updated, but I removed this function anyways since I don't use it

File: src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp
Comment (jared-paul) [reply src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp]: I think we determined we can't because only the world_ptr get's updated

File: src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.h
Comment (jared-paul) [reply src/software/simulated_tests/validation_functions/ball_in_rectangle_validation.h]: resolved

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (jared-paul) [reply src/software/simulated_tests/corner_kick_play_test.cpp]: resolved

Comment (jared-paul) [reply src/software/simulated_tests/corner_kick_play_test.cpp]: added a question to this PR about this

File: src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp
Comment (jared-paul) [reply src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp]: resolved

**Comment by jared-paul** [comment]
Should I add a check to see if the robot kicked the ball at the goal?

File: src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_received_ball_validation.cpp]: Something descriptive should be piped to this log. 
`LOG(FATAL) << "oh no it's broken :( ;`

File: src/software/ai/evaluation/robot.h
Comment (MathewMacDougall) [inline src/software/ai/evaluation/robot.h]: nit: missing newline at EOF

File: src/software/ai/evaluation/robot_test.cpp
Comment (MathewMacDougall) [inline src/software/ai/evaluation/robot_test.cpp]: nit: missing newline at EOF

File: src/software/simulated_tests/validation_functions/friendly_scored_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/friendly_scored_validation.cpp]: nit: missing newline at EOF

File: src/software/simulated_tests/validation_functions/friendly_scored_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/friendly_scored_validation.h]: nit: missing newline at EOF

File: src/software/simulated_tests/validation_functions/robot_received_ball_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_received_ball_validation.h]: nit: missing newline at EOF

File: src/software/ai/evaluation/robot.h
Comment (MathewMacDougall) [inline src/software/ai/evaluation/robot.h]: function should have javadoc

Also I find the name misleading compared to what the function seems to be doing. Based on the name I assume it checks if the given point is within some distance of the robot. However the function is actually checking for possession. Would be good to have a more descriptive name like `ballNearRobotMouth`, or `robotHasPossessionOfBall` if it's supposed to be for possession.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/ai/evaluation/robot.h
Comment (jonathanlew) [reply src/software/ai/evaluation/robot.h]: this was originally pointNearDribbler in the other PR, would that be better?

**Review by jonathanlew — COMMENTED** [review]
LGTM other than other reviewers' comments

File: src/software/ai/evaluation/robot.h
Comment (akhilveeraghanta) [reply src/software/ai/evaluation/robot.h]: This function is identical to the `isNearDribbler` function in #1768, depending on which goes PR goes in first we should probably use one or the other

**Review by akhilveeraghanta — COMMENTED** [review]
nice to see this coming together! not much to add beyond existing comments

File: src/software/ai/evaluation/robot.h
Comment (amrm3lm) [inline src/software/ai/evaluation/robot.h]: think its better for this to be `pointNearRobot(const Point &point, const Robot &robot)`

File: src/software/ai/evaluation/robot_test.cpp
Comment (amrm3lm) [inline src/software/ai/evaluation/robot_test.cpp]: duplicate line

**Review by amrm3lm — COMMENTED** [review]
nice stuff, not much to add beyond the other comments

File: src/software/ai/evaluation/robot.h
Comment (jared-paul) [reply src/software/ai/evaluation/robot.h]: resolved

File: src/software/ai/evaluation/robot_test.cpp
Comment (jared-paul) [reply src/software/ai/evaluation/robot_test.cpp]: resolved

File: src/software/ai/evaluation/robot.h
Comment (MathewMacDougall) [reply src/software/ai/evaluation/robot.h]: yea pointNearDribbler would also be better

**Review by MathewMacDougall — COMMENTED** [review]
Looks like a bunch of firmware, travis, and CI changes got included in this diff. Holding off a re-review until those disappear (should just need a merge).

**Review by MathewMacDougall — APPROVED** [review]
LGTM

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**
